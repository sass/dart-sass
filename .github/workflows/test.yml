name: Test

defaults:
  run: {shell: bash}

on:
  workflow_call:
  workflow_dispatch:

jobs:
  format:
    name: Code formatting
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v4
    - uses: dart-lang/setup-dart@v1
    - run: dart format --fix .
    - run: git diff --exit-code

  static_analysis:
    name: Static analysis
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with: {github-token: "${{ github.token }}"}

      - name: Analyze Dart
        run: dart analyze --fatal-warnings ./

  dartdoc:
    name: Dartdoc
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with: {github-token: "${{ github.token }}"}

      - name: dartdoc sass
        run: dart run dartdoc --quiet --no-generate-docs
          --errors ambiguous-doc-reference,broken-link,deprecated
          --errors unknown-directive,unknown-macro,unresolved-doc-reference
      - name: dartdoc sass_api
        run: cd pkg/sass_api && dart run dartdoc --quiet --no-generate-docs
          --errors ambiguous-doc-reference,broken-link,deprecated
          --errors unknown-directive,unknown-macro,unresolved-doc-reference

  sass_spec_language:
    name: "Language Tests | Dart ${{ matrix.dart_channel }} | ${{ matrix.async_label }}"
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        dart_channel: [stable, dev]
        async_label: [synchronous]
        async_args: ['']
        include:
          - dart_channel: stable
            async_label: asynchronous
            async_args: '--cmd-args --async'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
      - uses: ./.github/util/sass-spec

      - name: Run specs
        run: npm run sass-spec -- --dart .. $extra_args
        working-directory: sass-spec
        env: {extra_args: "${{ matrix.async_args }}"}

  sass_spec_js:
    name: "JS API Tests | Pure JS | Dart ${{ matrix.dart_channel }} | Node ${{ matrix.node-version }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3
          # Test LTS version with dart dev channel
          - os: ubuntu-latest
            dart_channel: dev
            node-version: 'lts/*'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}
      - uses: ./.github/util/sass-spec

      - name: Build JS
        run: dart run grinder pkg-npm-dev

      - name: Check out Sass specification
        uses: sass/clone-linked-repo@v1
        with:
          repo: sass/sass
          path: language

      - name: Run tests
        run: npm run js-api-spec -- --sassSassRepo ../language --sassPackage ../build/npm
        working-directory: sass-spec

  sass_spec_js_embedded:
    name: 'JS API Tests | Embedded | Node ${{ matrix.node-version }} | ${{ matrix.os }}'
    runs-on: ${{ matrix.os }}
    if: "github.event_name != 'pull_request' || !contains(github.event.pull_request.body, 'skip sass-embedded')"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}
      - uses: ./.github/util/sass-spec

      - name: Check out the embedded host
        uses: sass/clone-linked-repo@v1
        with: {repo: sass/embedded-host-node}

      - name: Check out the language repo
        uses: sass/clone-linked-repo@v1
        with: {repo: sass/sass, path: build/language}

      - name: Initialize embedded host
        run: |
          npm install
          npm run init -- --compiler-path=.. --language-path=../build/language
          npm run compile
          mv {`pwd`/,dist/}lib/src/vendor/dart-sass
        working-directory: embedded-host-node

      - name: Version info
        run: |
          path=embedded-host-node/dist/lib/src/vendor/dart-sass/sass
          if [[ -f "$path.cmd" ]]; then "./$path.cmd" --version
          elif [[ -f "$path.bat" ]]; then "./$path.bat" --version
          elif [[ -f "$path.exe" ]]; then "./$path.exe" --version
          else "./$path" --version
          fi

      - name: Run tests
        run: npm run js-api-spec -- --sassPackage ../embedded-host-node --sassSassRepo ../build/language
        working-directory: sass-spec

  sass_spec_js_browser:
    name: "JS API Tests | Browser | Dart ${{ matrix.dart_channel }}"

    strategy:
      matrix:
        dart_channel: [stable]
      fail-fast: false

    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: browser-actions/setup-chrome@v1
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
      - uses: ./.github/util/sass-spec

      - name: Build JS
        run: dart run grinder pkg-npm-dev

      - name: Install built dependencies
        run: npm install
        working-directory: build/npm

      - name: Check out Sass specification
        uses: sass/clone-linked-repo@v1
        with:
          repo: sass/sass
          path: language

      - name: Run tests
        run: npm run js-api-spec -- --sassSassRepo ../language --sassPackage ../build/npm --browser
        working-directory: sass-spec
        env:
          CHROME_EXECUTABLE: chrome

  dart_tests:
    name: "Dart tests | Dart ${{ matrix.dart_channel }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        # TODO(nweiz): Re-enable this when
        # https://github.com/dart-lang/sdk/issues/52121#issuecomment-1728534228
        # is addressed.
        # include: [{os: ubuntu-latest, dart_channel: dev}]

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}

      - run: dart run grinder pkg-standalone-dev
      - name: Run tests
        run: dart run test -x node

  # Unit tests that use Node.js, defined in test/.
  node_tests:
    name: "Node tests | Dart ${{ matrix.dart_channel }} | Node ${{ matrix.node-version }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false

      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3
          # Test LTS version with dart dev channel
          - os: ubuntu-latest
            dart_channel: dev
            node-version: 'lts/*'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}

      - run: dart run grinder pkg-npm-dev
      - name: Run tests
        run: dart run test -t node -j 2

  browser_tests:
    name: "Browser Tests | Dart ${{ matrix.dart_channel }}"

    strategy:
      matrix:
        dart_channel: [stable]
      fail-fast: false

    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - uses: browser-actions/setup-chrome@v1
    - uses: ./.github/util/initialize
      with:
        dart-sdk: ${{ matrix.dart_channel }}
        github-token: ${{ github.token }}

    - run: dart run grinder pkg-npm-dev
    - name: Run tests
      run: dart run test -p chrome -j 2
      env:
        CHROME_EXECUTABLE: chrome
