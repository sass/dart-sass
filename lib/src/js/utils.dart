// Copyright 2017 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:js_core/js_core.dart';
import 'package:node_interop/module.dart';
import 'package:node_interop/node_interop.dart';

import '../syntax.dart';
import '../value.dart';

/// Converts a JavaScript separator string into a [ListSeparator].
ListSeparator parseSeparator(String? separator) => switch (separator) {
      ' ' => ListSeparator.space,
      ',' => ListSeparator.comma,
      '/' => ListSeparator.slash,
      null => ListSeparator.undecided,
      _ => JSError.throwLikeJS(JSError('Unknown separator "$separator".')),
    };

/// Converts a syntax string to an instance of [Syntax].
Syntax parseSyntax(String? syntax) => switch (syntax) {
      null || 'scss' => Syntax.scss,
      'indented' => Syntax.sass,
      'css' => Syntax.css,
      _ => JSError.throwLikeJS(JSError('Unknown syntax "$syntax".')),
    };

/// The path to the Node.js entrypoint, if one can be located.
String? get entrypointFilename {
  if (requireNamespace.main?.filename case var filename?) {
    return filename;
  } else if (process.arguments case [_, String path, ...]) {
    return moduleModule.createRequireNamespace(path).resolve(path);
  } else {
    return null;
  }
}
