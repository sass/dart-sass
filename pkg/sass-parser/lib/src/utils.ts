// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import * as postcss from 'postcss';

import {Node} from './node';

/**
 * A type that matches any constructor for {@link T}. From
 * https://www.typescriptlang.org/docs/handbook/mixins.html.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Constructor<T> = new (...args: any[]) => T;

/**
 * An explicit field description passed to `cloneNode` that describes in detail
 * how to clone it.
 */
interface ExplicitClonableField<Name extends string> {
  /** The field's name. */
  name: Name;

  /**
   * Whether the field can be set to an explicit undefined value which means
   * something different than an absent field.
   */
  explicitUndefined?: boolean;
}

/** The type of field names that can be passed into `cloneNode`. */
type ClonableField<Name extends string> = Name | ExplicitClonableField<Name>;

/** Makes a {@link ClonableField} explicit. */
function parseClonableField<Name extends string>(
  field: ClonableField<Name>
): ExplicitClonableField<Name> {
  return typeof field === 'string' ? {name: field} : field;
}

/**
 * Creates a copy of {@link node} by passing all the properties in {@link
 * constructorFields} as an object to its constructor.
 *
 * If {@link overrides} is passed, it overrides any existing constructor field
 * values. It's also used to assign {@link assignedFields} after the cloned
 * object has been constructed.
 */
export function cloneNode<T extends Pick<postcss.Node, 'source'>>(
  node: T,
  overrides: Record<string, unknown> | undefined,
  constructorFields: ClonableField<keyof T & string>[],
  assignedFields?: ClonableField<keyof T & string>[]
): T {
  // We have to do these casts because the actual `...Prop` types that get
  // passed in and used for the constructor aren't actually subtypes of
  // `Partial<T>`. They use `never` types to ensure that various properties are
  // mutually exclusive, which is not compatible.
  const typedOverrides = overrides as Partial<T> | undefined;
  const constructorFn = node.constructor as new (defaults: Partial<T>) => T;

  const constructorParams: Partial<T> = {};
  for (const field of constructorFields) {
    const {name, explicitUndefined} = parseClonableField(field);
    let value: T[keyof T & string] | undefined;
    if (
      typedOverrides &&
      (explicitUndefined
        ? Object.hasOwn(typedOverrides, name)
        : typedOverrides[name] !== undefined)
    ) {
      value = typedOverrides[name];
    } else {
      value = maybeClone(node[name]);
    }
    if (value !== undefined) constructorParams[name] = value;
  }
  const cloned = new constructorFn(constructorParams);

  if (typedOverrides && assignedFields) {
    for (const field of assignedFields) {
      const {name, explicitUndefined} = parseClonableField(field);
      if (
        explicitUndefined
          ? Object.hasOwn(typedOverrides, name)
          : typedOverrides[name]
      ) {
        // This isn't actually guaranteed to be non-null, but TypeScript
        // (correctly) complains that we could be passing an undefined value to
        // a field that doesn't allow undefined. We don't have a good way of
        // forbidding that while still allowing users to override values that do
        // explicitly allow undefined, though.
        cloned[name] = typedOverrides[name]!;
      }
    }
  }

  cloned.source = node.source;
  return cloned;
}

/**
 * If {@link value} is a Sass node, a record, or an array, clones it and returns
 * the clone. Otherwise, returns it as-is.
 */
function maybeClone<T>(value: T): T {
  if (Array.isArray(value)) return value.map(maybeClone) as T;
  if (typeof value !== 'object' || value === null) return value;
  // The only records we care about are raws, which only contain primitives and
  // arrays of primitives, so structued cloning is safe.
  if (value.constructor === Object) return structuredClone(value);
  if (value instanceof postcss.Node) return value.clone() as T;
  return value;
}

/**
 * Converts {@link node} into a JSON-safe object, with the given {@link fields}
 * included.
 *
 * This always includes the `type`, `sassType`, `raws`, and `source` fields if
 * set. It converts multiple references to the same source input object into
 * indexes into a top-level list.
 */
export function toJSON<T extends Node>(
  node: T,
  fields: (keyof T & string)[],
  inputs?: Map<postcss.Input, number>
): object {
  // Only include the inputs field at the top level.
  const includeInputs = !inputs;
  inputs ??= new Map();
  let inputIndex = inputs.size;

  const result: Record<string, unknown> = {};
  if ('type' in node) result.type = (node as {type: string}).type;

  fields = ['sassType', 'raws', ...fields];
  for (const field of fields) {
    const value = node[field];
    if (value !== undefined) result[field] = toJsonField(field, value, inputs);
  }

  if (node.source) {
    let inputId = inputs.get(node.source.input);
    if (inputId === undefined) {
      inputId = inputIndex++;
      inputs.set(node.source.input, inputId);
    }

    result.source = {
      start: node.source.start,
      end: node.source.end,
      inputId,
    };
  }

  if (includeInputs) {
    result.inputs = [...inputs.keys()].map(input => input.toJSON());
  }
  return result;
}

/**
 * Converts a single field with name {@link field} and value {@link value} to a
 * JSON-safe object.
 *
 * The {@link inputs} map works the same as it does in {@link toJSON}.
 */
function toJsonField(
  field: string,
  value: unknown,
  inputs: Map<postcss.Input, number>
): unknown {
  if (typeof value !== 'object' || value === null) {
    return value;
  } else if (Array.isArray(value)) {
    return value.map((element, i) =>
      toJsonField(i.toString(), element, inputs)
    );
  } else if ('toJSON' in value) {
    if ('sassType' in value) {
      return (
        value as {
          toJSON: (field: string, inputs: Map<postcss.Input, number>) => object;
        }
      ).toJSON('', inputs);
    } else {
      return (value as {toJSON: (field: string) => object}).toJSON(field);
    }
  } else {
    return value;
  }
}

/**
 * Returns the longest string (of code units) that's an initial substring of
 * every string in
 * {@link strings}.
 */
export function longestCommonInitialSubstring(strings: string[]): string {
  let candidate: string | undefined;
  for (const string of strings) {
    if (candidate === undefined) {
      candidate = string;
    } else {
      for (let i = 0; i < candidate.length && i < string.length; i++) {
        if (candidate.charCodeAt(i) !== string.charCodeAt(i)) {
          candidate = candidate.substring(0, i);
          break;
        }
      }
      candidate = candidate.substring(
        0,
        Math.min(candidate.length, string.length)
      );
    }
  }
  return candidate ?? '';
}
