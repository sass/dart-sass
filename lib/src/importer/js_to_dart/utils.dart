// Copyright 2023 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:node_interop/js.dart';

import '../../js/utils.dart';
import '../utils.dart';

/// Throws a JsError if [scheme] isn't a valid URL scheme.
void validateUrlScheme(String scheme) {
  if (!isValidUrlScheme(scheme)) {
    jsThrow(
        JsError('"$scheme" isn\'t a valid URL scheme (for example "file").'));
  }
}
