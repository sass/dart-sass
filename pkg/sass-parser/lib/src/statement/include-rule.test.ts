// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import {
  ArgumentList,
  GenericAtRule,
  IncludeRule,
  ParameterList,
  sass,
  scss,
} from '../..';
import * as utils from '../../../test/utils';

describe('a @include rule', () => {
  let node: IncludeRule;
  describe('with no block', () => {
    function describeNode(
      description: string,
      create: () => IncludeRule,
    ): void {
      describe(description, () => {
        beforeEach(() => void (node = create()));

        it('has a sassType', () => expect(node.sassType).toBe('include-rule'));

        it('has a name', () => expect(node.name.toString()).toBe('include'));

        it('has an include name', () =>
          expect(node.includeName.toString()).toBe('foo'));

        it('has an argument', () =>
          expect(node.arguments.nodes[0]).toHaveStringExpression(
            'value',
            'bar',
          ));

        it('has no using', () => expect(node.using).toBe(undefined));

        it('has matching params', () => expect(node.params).toBe('foo(bar)'));

        it('has no nodes', () => expect(node.nodes).toBeUndefined());
      });
    }

    describeNode(
      'parsed as SCSS',
      () => scss.parse('@include foo(bar)').nodes[0] as IncludeRule,
    );

    describeNode(
      'parsed as Sass',
      () => sass.parse('@include foo(bar)').nodes[0] as IncludeRule,
    );

    describeNode(
      'constructed manually',
      () => new IncludeRule({includeName: 'foo', arguments: [{text: 'bar'}]}),
    );

    describeNode('constructed from ChildProps', () =>
      utils.fromChildProps({includeName: 'foo', arguments: [{text: 'bar'}]}),
    );
  });

  describe('with a child', () => {
    function describeNode(
      description: string,
      create: () => IncludeRule,
    ): void {
      describe(description, () => {
        beforeEach(() => void (node = create()));

        it('has a sassType', () => expect(node.sassType).toBe('include-rule'));

        it('has a name', () => expect(node.name.toString()).toBe('include'));

        it('has a include name', () =>
          expect(node.includeName.toString()).toBe('foo'));

        it('has an argument', () =>
          expect(node.arguments.nodes[0]).toHaveStringExpression(
            'value',
            'bar',
          ));

        it('has no using', () => expect(node.using).toBe(undefined));

        it('has matching params', () => expect(node.params).toBe('foo(bar)'));

        it('has a child node', () => {
          expect(node.nodes).toHaveLength(1);
          expect(node.nodes![0]).toBeInstanceOf(GenericAtRule);
          expect(node.nodes![0]).toHaveInterpolation(
            'nameInterpolation',
            'baz',
          );
        });
      });
    }

    describeNode(
      'parsed as SCSS',
      () => scss.parse('@include foo(bar) {@baz}').nodes[0] as IncludeRule,
    );

    describeNode(
      'parsed as Sass',
      () => sass.parse('@include foo(bar)\n  @baz').nodes[0] as IncludeRule,
    );

    describeNode(
      'constructed manually',
      () =>
        new IncludeRule({
          includeName: 'foo',
          arguments: [{text: 'bar'}],
          nodes: [{name: 'baz'}],
        }),
    );

    describeNode('constructed from ChildProps', () =>
      utils.fromChildProps({
        includeName: 'foo',
        arguments: [{text: 'bar'}],
        nodes: [{name: 'baz'}],
      }),
    );
  });

  describe('with using', () => {
    function describeNode(
      description: string,
      create: () => IncludeRule,
    ): void {
      describe(description, () => {
        beforeEach(() => void (node = create()));

        it('has a sassType', () => expect(node.sassType).toBe('include-rule'));

        it('has a name', () => expect(node.name.toString()).toBe('include'));

        it('has a include name', () =>
          expect(node.includeName.toString()).toBe('foo'));

        it('has an argument', () =>
          expect(node.arguments.nodes[0]).toHaveStringExpression(
            'value',
            'bar',
          ));

        it('has a using parameter', () =>
          expect(node.using!.nodes[0].name).toBe('baz'));

        it('has matching params', () =>
          expect(node.params).toBe('foo(bar) using ($baz)'));

        it('has a child node', () => {
          expect(node.nodes).toHaveLength(1);
          expect(node.nodes![0]).toBeInstanceOf(GenericAtRule);
          expect(node.nodes![0]).toHaveInterpolation(
            'nameInterpolation',
            'qux',
          );
        });
      });
    }

    describeNode(
      'parsed as SCSS',
      () =>
        scss.parse('@include foo(bar) using ($baz) {@qux}')
          .nodes[0] as IncludeRule,
    );

    describeNode(
      'parsed as Sass',
      () =>
        sass.parse('@include foo(bar) using ($baz)\n  @qux')
          .nodes[0] as IncludeRule,
    );

    describeNode(
      'constructed manually',
      () =>
        new IncludeRule({
          includeName: 'foo',
          arguments: [{text: 'bar'}],
          using: ['baz'],
          nodes: [{name: 'qux'}],
        }),
    );

    describeNode('constructed from ChildProps', () =>
      utils.fromChildProps({
        includeName: 'foo',
        arguments: [{text: 'bar'}],
        using: ['baz'],
        nodes: [{name: 'qux'}],
      }),
    );
  });

  describe('throws an error when assigned a new', () => {
    beforeEach(
      () =>
        void (node = scss.parse('@include foo(bar)').nodes[0] as IncludeRule),
    );

    it('name', () => expect(() => (node.name = 'qux')).toThrow());

    it('params', () => expect(() => (node.params = 'zip(zap)')).toThrow());
  });

  describe('assigned new arguments', () => {
    beforeEach(
      () =>
        void (node = scss.parse('@include foo(bar)').nodes[0] as IncludeRule),
    );

    it("removes the old arguments' parent", () => {
      const oldArguments = node.arguments;
      node.arguments = [{text: 'qux'}];
      expect(oldArguments.parent).toBeUndefined();
    });

    it("assigns the new arguments' parent", () => {
      const args = new ArgumentList([{text: 'qux'}]);
      node.arguments = args;
      expect(args.parent).toBe(node);
    });

    it('assigns the arguments explicitly', () => {
      const args = new ArgumentList([{text: 'qux'}]);
      node.arguments = args;
      expect(node.arguments).toBe(args);
    });

    it('assigns the expression as ArgumentProps', () => {
      node.arguments = [{text: 'qux'}];
      expect(node.arguments.nodes[0]).toHaveStringExpression('value', 'qux');
      expect(node.arguments.parent).toBe(node);
    });
  });

  describe('assigned new using', () => {
    beforeEach(
      () =>
        void (node = scss.parse('@include foo(bar) using ($baz) {}')
          .nodes[0] as IncludeRule),
    );

    it("removes the old using' parent", () => {
      const oldUsing = node.using!;
      node.using = ['qux'];
      expect(oldUsing.parent).toBeUndefined();
    });

    it("assigns the new using' parent", () => {
      const using = new ParameterList(['qux']);
      node.using = using;
      expect(using.parent).toBe(node);
    });

    it('assigns the using explicitly', () => {
      const using = new ParameterList(['qux']);
      node.using = using;
      expect(node.using).toBe(using);
    });

    it('assigns the expression as ParameterProps', () => {
      node.using = ['qux'];
      expect(node.using!.nodes[0].name).toBe('qux');
      expect(node.using!.parent).toBe(node);
    });
  });

  describe('stringifies', () => {
    describe('to SCSS', () => {
      describe('with default raws', () => {
        it('with no arguments', () =>
          expect(new IncludeRule({includeName: 'foo'}).toString()).toBe(
            '@include foo',
          ));

        it('with an argument', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              arguments: [{text: 'bar'}],
            }).toString(),
          ).toBe('@include foo(bar)'));

        it('with empty using', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              using: [],
              nodes: [],
            }).toString(),
          ).toBe('@include foo using () {}'));

        it('with a using parameter', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              using: ['bar'],
              nodes: [],
            }).toString(),
          ).toBe('@include foo using ($bar) {}'));

        it('with a non-identifier name', () =>
          expect(
            new IncludeRule({
              includeName: 'f o',
              arguments: [{text: 'bar'}],
            }).toString(),
          ).toBe('@include f\\20o(bar)'));
      });

      it('with afterName', () =>
        expect(
          new IncludeRule({
            includeName: 'foo',
            raws: {afterName: '/**/'},
          }).toString(),
        ).toBe('@include/**/foo'));

      it('with matching includeName', () =>
        expect(
          new IncludeRule({
            includeName: 'foo',
            raws: {includeName: {value: 'foo', raw: 'f\\6fo'}},
          }).toString(),
        ).toBe('@include f\\6fo'));

      it('with non-matching includeName', () =>
        expect(
          new IncludeRule({
            includeName: 'foo',
            raws: {includeName: {value: 'fao', raw: 'f\\41o'}},
          }).toString(),
        ).toBe('@include foo'));

      it('with showArguments = true', () =>
        expect(
          new IncludeRule({
            includeName: 'foo',
            raws: {showArguments: true},
          }).toString(),
        ).toBe('@include foo()'));

      it('ignores showArguments with an argument', () =>
        expect(
          new IncludeRule({
            includeName: 'foo',
            arguments: [{text: 'bar'}],
            raws: {showArguments: false},
          }).toString(),
        ).toBe('@include foo(bar)'));

      describe('with afterArguments', () => {
        it('with no using', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              arguments: [{text: 'bar'}],
              raws: {afterArguments: '/**/'},
            }).toString(),
          ).toBe('@include foo(bar)'));

        it('with using', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              arguments: [{text: 'bar'}],
              using: ['baz'],
              nodes: [],
              raws: {afterArguments: '/**/'},
            }).toString(),
          ).toBe('@include foo(bar)/**/using ($baz) {}'));

        it('with no arguments', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              using: ['baz'],
              raws: {afterArguments: '/**/'},
            }).toString(),
          ).toBe('@include foo/**/using ($baz)'));
      });

      describe('with afterUsing', () => {
        it('with no using', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              arguments: [{text: 'bar'}],
              raws: {afterUsing: '/**/'},
            }).toString(),
          ).toBe('@include foo(bar)'));

        it('with using', () =>
          expect(
            new IncludeRule({
              includeName: 'foo',
              using: ['baz'],
              raws: {afterUsing: '/**/'},
            }).toString(),
          ).toBe('@include foo using/**/($baz)'));
      });
    });
  });

  describe('clone', () => {
    let original: IncludeRule;
    beforeEach(() => {
      original = scss.parse('@include foo(bar) using ($baz) {}')
        .nodes[0] as IncludeRule;
      // TODO: remove this once raws are properly parsed
      original.raws.between = '  ';
    });

    describe('with no overrides', () => {
      let clone: IncludeRule;
      beforeEach(() => void (clone = original.clone()));

      describe('has the same properties:', () => {
        it('params', () => expect(clone.params).toBe('foo(bar) using ($baz)'));

        it('includeName', () => expect(clone.includeName).toBe('foo'));

        it('arguments', () => {
          expect(clone.arguments.nodes[0]).toHaveStringExpression(
            'value',
            'bar',
          );
          expect(clone.arguments.parent).toBe(clone);
        });

        it('using', () => {
          expect(clone.using!.nodes[0].name).toBe('baz');
          expect(clone.using!.parent).toBe(clone);
        });

        it('raws', () => expect(clone.raws).toEqual({between: '  '}));

        it('source', () => expect(clone.source).toBe(original.source));
      });

      describe('creates a new', () => {
        it('self', () => expect(clone).not.toBe(original));

        for (const attr of ['arguments', 'using', 'raws'] as const) {
          it(attr, () => expect(clone[attr]).not.toBe(original[attr]));
        }
      });
    });

    describe('overrides', () => {
      describe('raws', () => {
        it('defined', () =>
          expect(original.clone({raws: {afterName: '  '}}).raws).toEqual({
            afterName: '  ',
          }));

        it('undefined', () =>
          expect(original.clone({raws: undefined}).raws).toEqual({
            between: '  ',
          }));
      });

      describe('includeName', () => {
        describe('defined', () => {
          let clone: IncludeRule;
          beforeEach(() => {
            clone = original.clone({includeName: 'qux'});
          });

          it('changes params', () =>
            expect(clone.params).toBe('qux(bar) using ($baz)'));

          it('changes includeName', () =>
            expect(clone.includeName).toEqual('qux'));
        });

        describe('undefined', () => {
          let clone: IncludeRule;
          beforeEach(() => {
            clone = original.clone({includeName: undefined});
          });

          it('preserves params', () =>
            expect(clone.params).toBe('foo(bar) using ($baz)'));

          it('preserves includeName', () =>
            expect(clone.includeName).toEqual('foo'));
        });
      });

      describe('arguments', () => {
        describe('defined', () => {
          let clone: IncludeRule;
          beforeEach(() => {
            clone = original.clone({arguments: [{text: 'qux'}]});
          });

          it('changes params', () =>
            expect(clone.params).toBe('foo(qux) using ($baz)'));

          it('changes arguments', () => {
            expect(clone.arguments.nodes[0]).toHaveStringExpression(
              'value',
              'qux',
            );
            expect(clone.arguments.parent).toBe(clone);
          });
        });

        describe('undefined', () => {
          let clone: IncludeRule;
          beforeEach(() => {
            clone = original.clone({arguments: undefined});
          });

          it('preserves params', () =>
            expect(clone.params).toBe('foo(bar) using ($baz)'));

          it('preserves arguments', () =>
            expect(clone.arguments.nodes[0]).toHaveStringExpression(
              'value',
              'bar',
            ));
        });
      });

      describe('using', () => {
        describe('defined', () => {
          let clone: IncludeRule;
          beforeEach(() => {
            clone = original.clone({using: ['qux']});
          });

          it('changes params', () =>
            expect(clone.params).toBe('foo(bar) using ($qux)'));

          it('changes arguments', () => {
            expect(clone.using!.nodes[0].name).toBe('qux');
            expect(clone.using!.parent).toBe(clone);
          });
        });

        describe('undefined', () => {
          let clone: IncludeRule;
          beforeEach(() => {
            clone = original.clone({using: undefined});
          });

          it('changes params', () => expect(clone.params).toBe('foo(bar)'));

          it('changes using', () => expect(clone.using).toBeUndefined());
        });
      });
    });
  });

  describe('toJSON', () => {
    it('with no children', () =>
      expect(scss.parse('@include foo(bar)').nodes[0]).toMatchSnapshot());

    it('with a child', () =>
      expect(
        scss.parse('@include foo(bar) {@qux}').nodes[0],
      ).toMatchSnapshot());

    it('with using and a child', () =>
      expect(
        scss.parse('@include foo(bar) using ($baz) {@qux}').nodes[0],
      ).toMatchSnapshot());
  });
});
