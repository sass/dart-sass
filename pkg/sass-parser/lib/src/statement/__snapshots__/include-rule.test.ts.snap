// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a @include rule toJSON with a child 1`] = `
{
  "arguments": <(bar)>,
  "includeName": "foo",
  "inputs": [
    {
      "css": "@include foo(bar) {@qux}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "include",
  "nodes": [
    <@qux>,
  ],
  "params": "foo(bar)",
  "raws": {},
  "sassType": "include-rule",
  "source": <1:1-1:25 in 0>,
  "type": "atrule",
}
`;

exports[`a @include rule toJSON with no children 1`] = `
{
  "arguments": <(bar)>,
  "includeName": "foo",
  "inputs": [
    {
      "css": "@include foo(bar)",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "include",
  "params": "foo(bar)",
  "raws": {},
  "sassType": "include-rule",
  "source": <1:1-1:18 in 0>,
  "type": "atrule",
}
`;

exports[`a @include rule toJSON with using and a child 1`] = `
{
  "arguments": <(bar)>,
  "includeName": "foo",
  "inputs": [
    {
      "css": "@include foo(bar) using ($baz) {@qux}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "include",
  "nodes": [
    <@qux>,
  ],
  "params": "foo(bar) using ($baz)",
  "raws": {},
  "sassType": "include-rule",
  "source": <1:1-1:38 in 0>,
  "type": "atrule",
  "using": <($baz)>,
}
`;
